### PR TITLE
Add AdditiveFOAM package

### PR DESCRIPTION
**Pull request by streeve**
Add initial AdditiveFOAM package, based on OpenFOAM-10

### CODE REVIEW

**Comment by streeve** [comment]
@gknapp1 do you have time to look at what the install function should look like here?

**Comment by streeve** [comment]
@kjrstory we have an OpenFOAM-based app and are trying to work out the best way to spack install. Right now we have a placeholder manual `install` step, but I think it should either rely on `OpenfoamOrg` package details or become an `OpenfoamOrgPackage` (due to `wmake`). Any initial thoughts?

**Comment by kjrstory** [comment]
It might be a good idea to reference the "of-precice" package for guidance on making these changes. To do this, you would need a separate Allwmake file. It's advisable to avoid using os.system if possible.

**Comment by kjrstory** [comment]
@streeve 
I have recreated it with reference to "of-precise." The main idea is to use the "spack-derived-Allwmake" of the OpenFOAM package to create the configure, build, and install steps.
```python
# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
# Spack Project Developers. See the top-level COPYRIGHT file for details.
#
# SPDX-License-Identifier: (Apache-2.0 OR MIT)

import os

from spack.package import *
from spack.pkg.builtin.openfoam import add_extra_files

class Additivefoam(Package):
    """AdditiveFOAM is a heat and mass transfer software for Additive Manufacturing (AM)"""

    homepage = "https://github.com/ORNL/AdditiveFOAM"
    git = "https://github.com/ORNL/AdditiveFOAM.git"
    url = "https://github.com/ORNL/AdditiveFOAM/archive/1.0.0.tar.gz"

    maintainers("streeve", "colemanjs", "gknapp1")

    tags = ["ecp"]

    version("main", branch="main")
    version("1.0.0", sha256="abbdf1b0230cd2f26f526be76e973f508978611f404fe8ec4ecdd7d5df88724c")

    depends_on("openfoam-org@10")

    common = ["spack-derived-Allwmake"]
    assets = ["applications/Allwmake", "Allwmake"]

    build_script = "./spack-derived-Allwmake"

    phases = ['configure', 'build', 'install']

    def patch(self):
        add_extra_files(self, self.common, self.assets)

    def configure(self, spec, prefix):
        pass

    def build(self, spec, prefix):
        """Build with Allwmake script, wrapped to source environment first."""
        args = []
        if self.parallel:  # Parallel build? - pass via environment
           os.environ['WM_NCOMPPROCS'] = str(make_jobs)
        builder = Executable(self.build_script)
        builder(*args)


    def install(self, spec, prefix):
        """Install under the prefix directory"""

        for f in ['README.md', 'LICENSE']:
            if os.path.isfile(f):
                install(f, join_path(self.prefix, f))

        dirs = ['tutorials', 'applications']
        for d in dirs:
            if os.path.isdir(d):
                install_tree(
                    d,
                    join_path(self.prefix, d),
                    symlinks=True)
```

assets/Allwmake
```
#!/bin/sh
cd ${0%/*} || exit 1    # Run from this directory

applications/Allwmake $targetType $*
```

assets/applications/Allwmake
```
#!/bin/sh
cd ${0%/*} || exit 1    # Run from this directory

wmake libso solvers/additiveFoam/movingHeatSource
wmake solvers/additiveFoam

```

When I tested it, everything worked fine. I would appreciate it if you consider accepting the suggestion.

**Comment by streeve** [comment]
Thanks very much for your help, @kjrstory. I pushed your changes - @gknapp1 can you test on your side as well?

**Comment by gknapp1** [comment]
I am getting a permission issue for the `Allwmake` script when I try to `spack install additivefoam`. See log file below.

Is there a spack function that can be added to the package to give executable permissions to `Allwmake`?

```
==> additivefoam: Executing phase: 'configure'
==> additivefoam: Executing phase: 'build'
==> [2023-09-18-14:59:41.235085] './spack-derived-Allwmake'
========================================
2023-09-18 14:59:41 +0000
Build with OpenFOAM-10
  WM_PROJECT_DIR = /home/cloud/spack_reeve/spack/opt/spack/linux-centos7-zen2/gcc-13.1.0/openfoam-org-10-4m5cuouqhiw3crmhshayst4ovipac422
  Gcc system compiler
  linux64GccDPInt32-spack - with SYSTEMMPI mpi-system

./spack-derived-Allwmake: line 31: ./Allwmake: Permission denied
```

**Comment by gknapp1** [comment]
> I am getting a permission issue for the `Allwmake` script when I try to `spack install additivefoam`. See log file below.
> ...

Turns out the executable permission was not set on the repository version of the Allwmake files. Fixed this and it installs fine for me now.

**Review by gknapp1 — APPROVED** [review]
I've been able to successfully install the `additivefoam` spack package using this, and it all looks good to me. Thanks for your help getting this working @kjrstory!

**Comment by streeve** [comment]
@colemanjs take a quick look and approve if you can

**Review by colemanjs — APPROVED** [review]
spack install of additivefoam successful on Ubuntu20.04

File: var/spack/repos/builtin/packages/additivefoam/package.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    url = "https://github.com/ORNL/AdditiveFOAM/archive/1.0.0.tar.gz"
+
+    maintainers("streeve", "colemanjs", "gknapp1")
+
+    tags = ["ecp"]
+
+    version("main", branch="main")
+    version("1.0.0", sha256="abbdf1b0230cd2f26f526be76e973f508978611f404fe8ec4ecdd7d5df88724c")
+
+    depends_on("openfoam-org@10")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/additivefoam/package.py:26]: Is this at v10 only or do you anticipate v10 on?

File: var/spack/repos/builtin/packages/additivefoam/package.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    def patch(self):
+        add_extra_files(self, self.common, self.assets)
+
+    def configure(self, spec, prefix):
+        pass
+
+    def build(self, spec, prefix):
+        """Build with Allwmake script, wrapped to source environment first."""
+        args = []
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/additivefoam/package.py:43]: Are there supposed to be any `args` at this point?

**Review by tldahlgren — COMMENTED** [review]
Confirmed the homepage, branch, and version sha256.

File: var/spack/repos/builtin/packages/additivefoam/package.py (hunk 1 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/additivefoam/package.py:26]: @colemanjs correct me if I'm wrong, but I believe we only intend to use v10 and are not compatible with v11

File: var/spack/repos/builtin/packages/additivefoam/package.py (hunk 2 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/additivefoam/package.py:43]: No current intended args; written to match the `of-precise` package and potentially extend

**Review by tldahlgren — APPROVED** [review]
I'm able to successfully built the package.

**Comment by tldahlgren** [comment]
Ping me when this PR is ready to be merged.

File: var/spack/repos/builtin/packages/additivefoam/package.py (hunk 1 above)
Comment (colemanjs) [reply var/spack/repos/builtin/packages/additivefoam/package.py:26]: @streeve, AdditiveFOAM is only compatible with v10 at the moment. We are currently porting to OpenFOAM-11, however backward compatibility is not guaranteed. I am not sure how best to handle this other than specifying which openfoam version to build for a given version of AdditiveFOAM. Thoughts/suggestions?

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/additivefoam/package.py:26]: Sounds like this PR is sufficient ATM and another PR can be submitted after the port.

**Pull request merged**